assignments/burmanm/week-5.scm

```diff
@@ -0,0 +1,583 @@
+;; Exercise 2.17
+
+(define (last-pair list)
+  (if (null? (cdr list))
+      list
+      (last-pair (cdr list))))
+
+
+;; scheme@(guile-user) [6]> (last-pair (list 23 72 149 34))
+;; $14 = (34)
+
+;; Exercise 2.18
+
+(define (reverse list)
+  (if (null? list)
+      '()
+      (append (reverse (cdr list)) (cons (car list) '()))))
+
+;; Exercise 2.19
+
+(define (except-first-denomination coin-values)
+  (cdr coin-values))
+
+(define (first-denomination coin-values)
+  (car coin-values))
+
+(define (no-more? coin-values)
+  (null? coin-values))
+
+(define (cc amount coin-values)
+  (cond ((= amount 0) 1)
+	((or (< amount 0) (no-more? coin-values)) 0)
+	(else
+	 (+ (cc amount
+		(except-first-denomination
+		 coin-values))
+	    (cc (- amount
+		   (first-denomination
+		    coin-values))
+		coin-values)))))
+
+;; (define us-coins (list 50 25 10 5 1))
+;; (define uk-coins (list 100 50 20 10 5 2 1 0.5))
+
+;; scheme@(guile-user) [6]> (cc 100 us-coins)
+;; $27 = 292
+;; scheme@(guile-user) [6]> (cc 100 uk-coins)
+;; $28 = 104561
+;; scheme@(guile-user) [6]> 
+
+;; It does make a difference, since larger amount of coins creates larger recursion tree and more potential answers.
+
+;; Exercise 2.20
+
+(define (same-parity x . y)
+  (define (not-even? z) (not (even? z)))
+  (define (print-list comp l1 l2)
+    (if (null? l2)
+	l1
+	(if (comp (car l2))
+	    (print-list comp (cons l1 (car l2)) (cdr l2))
+	    (print-list comp l1 (cdr l2)))))
+  (if (even? x)
+      (print-list even? (list x) y)
+      (print-list not-even? (list x) y)))
+
+;; Exercise 2.21
+
+;; For guile, not needed with mit scheme
+(define (square x) (* x x))
+
+;; nil is not known by mit-scheme or guile..
+(define nil '())
```

Comment:
yeah, it's explained in the [footnote 10](https://mitpress.mit.edu/sites/default/files/sicp/full-text/book/book-Z-H-15.html#footnote_Temp_158) :)